crates/cext/src/dag.rs

```diff
@@ -157,6 +163,509 @@ pub unsafe extern "C" fn qk_dag_num_clbits(dag: *const DAGCircuit) -> u32 {
     dag.num_clbits() as u32
 }
 
+/// @ingroup QkDag
+/// Return the total number of operation nodes in the DAG.
+///
+/// @param dag A pointer to the DAG.
+///
+/// @return The total number of instructions in the DAG.
+///
+/// # Example
+/// ```c
+///     QkDag *dag = qk_dag_new();
+///     QkQuantumRegister *qr = qk_quantum_register_new(1, "my_register");
+///     qk_dag_add_quantum_register(dag, qr);
+///
+///     uint32_t qubit[1] = {0};
+///     qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+///     size_t num = qk_dag_num_op_nodes(dag); // 1
+///
+///     qk_dag_free(dag);
+///     qk_quantum_register_free(qr);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_num_op_nodes(dag: *const DAGCircuit) -> usize {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    dag.num_ops()
+}
+
+/// @ingroup QkDag
+///
+/// DAG node type.
+#[repr(u8)]
+pub enum CDagNodeType {
```

Comment:
do we need/want to derive clone/copy/debug and friends on this while we're at it?